text/0066-separate-context-propagation.md

```diff
@@ -0,0 +1,499 @@
+
+**RemoveBaggage(context, key) -> context**  
+To delete distributed state from an aspect, the Baggage API provides a function 
+which takes a context, a key, and a value as input, and returns an updated 
+context which contains the new value.
+
+**ClearBaggage(context) -> context**  
+To avoid sending baggage to an untrusted downstream process, the Baggage API 
+provides a function remove all baggage from a context.
+
+**GetBaggageHTTPPropagator() -> (HTTPExtract, HTTPInject)**  
+To deserialize the state of the system sent from the the prior upstream process, 
+and to serialize the the current state of the system and send it to the next 
+downstream process, the Baggage API provides a function which returns a 
+baggage-specific implementation of the HTTPExtract and HTTPInject functions.
+
+# Context Propagation
+
+## Context API
+
+Aspects access data in-process using a shared context object. Each aspect uses 
+its own namespaced set of keys in the context, containing all of the data for 
+that cross-cutting concern.
+
+**SetValue(context, key, value) -> context**  
+To record the local state of an aspect, the Context API provides a function 
+which takes a context, a key, and a value as input, and returns an updated 
+context which contains the new value.
+
+**GetValue(context, key) -> value**  
+To access the local state of an aspect, the Context API provides a function 
+which takes a context and a key as input, and returns a value.
+
+### Optional: Automated Context Management
+When possible, the OpenTelemetry context should automatically be associated 
+with the program execution context. Note that some languages do not provide any 
+facility for setting and getting a current context. In these cases, the user is 
+responsible for managing the current context.  
+
+**SetCurrent(context)**  
+To associate a context with program execution, the Context API provides a 
+function which takes a Context.
+
+**GetCurrent() -> context**  
+To access the context associated with program execution, the Context API 
+provides a function which takes no arguments and returns a Context.
+
+
+## Propagation API
+
+Aspects send their state to downstream processes via propagators: 
+functions which read and write context into RPC requests. Each aspect creates a 
+set of propagators for every type of supported medium - currently only HTTP 
+requests.
+
+**HTTPExtract(context, request) -> context**  
+To receive data injected by prior upstream processes, the Propagation API 
+provides a function which takes a context and an HTTP request, and returns 
+context which represents the state of the upstream system.
+
+**HTTPInject(context, request)**  
+To send the data for all aspects downstream to the next process, the 
+Propagation API provides a function which takes a context and an HTTP request, 
+and mutates the HTTP request to include an HTTP Header representation of the 
+context.
+
+**ChainHTTPInjector(injector, injector) -> injector**  
```

Comment:
I would argue strongly against this API, as it requires the fairly ugly (IMO) pattern of calling ChainHTTPInjector over and over again. Why not modify the APIs or expose a way to consume a list with implicit chaining?

effectively arguing for:

```
SetHTTPExtractors([BaggageExtractor(), TracingExtractor(), CorrelationExtractor()])
```

instead of:
```
SetHTTPExtractor(
    ChainHTTPExtractor(
        BaggageExtractor(),
        ChainHTTPExtractor(
            TracingExtractor(), 
            CorrelationExtractor()
        )
    )
)
```

Reply:
Using varargs is fine by me, as is abstracting away the chaining using a register pattern. However I would still argue that there needs to be some defined behavior around how the propagation will work with multiple propagators. But I'm open to putting that into a different PR or OTEP.

Reply:
I think you're referring to multiple propagators of the same type, e.g. propagator for trace context using W3C vs. B3 format. At this layer that's not a concern, the multiple propagators here are of different type, like baggage vs. correlations vs. trace context.

Reply:
I'm fine with not specifying that detail in this discussion. I'm +1 on varargs.